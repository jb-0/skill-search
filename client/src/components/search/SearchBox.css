.search-container {
  width: 400px;
  margin: auto;
}

.search-container-mobile {
  width: 80%;
}

#searchInputText {
  width: 80%;
  height: 40px;
  padding: 0 5%;
  color: #023047;
  font-size: 1.5rem;
  border-radius: 25px;
  outline-width: 0;
  border: solid;
  border-color: #023047;
}
