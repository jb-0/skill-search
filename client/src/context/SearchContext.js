/**
  * This context is used to pass state to the numerous components and events (such as save) that
  * make up a search. This also means a user can leave the search screen to view their profile
  * without losing search information.
  */
import React, { createContext, useState  } from 'react';

export const SearchContext = createContext();

export const SearchProvider = (props) => {
  // Store a number of variables in state, firstly the input from the search text box, second the
  // terms that have been added so far and finally the suggested search terms
  const [search, setSearch] = useState({
    searchInputText: '',
    searchTerms: [],
    suggestedTerms: [],
    submittedSearchTerms: [],
    location: 'london',
    submittedLocation: '',
    handleDropDownSelectUpdates,
    handleSearchButtonClick
  });

  function handleDropDownSelectUpdates(event) {
    const target = event.target.name;
    const value = event.target.value;

    setSearch((previousValues) => {
      return { ...previousValues, [target]: value };
    });
  }

  // When a user hits search ensure the location and searchTerms are current
  function handleSearchButtonClick() {
    setSearch((previousValues) => {
      return {
        ...previousValues,
        submittedSearchTerms: [...previousValues.searchTerms],
        submittedLocation: previousValues.location,
      };
    });
  }
  
  return <SearchContext.Provider value={ [search, setSearch] }>{props.children}</SearchContext.Provider>;
};
