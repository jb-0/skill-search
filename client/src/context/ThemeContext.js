/**
 * The theme context provides styling theme attributes as well as defining global styles
 */
import React, { createContext } from 'react';
import { createGlobalStyle } from 'styled-components';

export const ThemeContext = createContext();

export const ThemeProvider = (props) => {
  const theme = {
    primaryColor: '#212529',
    secondaryColor: 'white',
    tertiaryColor: '#ffb703',
    primaryPad: '20px',
    mainFont: `Nunito', sans-serif`,
  };

  return (
    <ThemeContext.Provider value={theme}>
      {props.children}
    </ThemeContext.Provider>
  );
};

export const GlobalStyle = createGlobalStyle`
  body {
  margin: 0;
  font-family: var(--main-font);
  text-align: center;
  color: var(--primary-color);
  background-color: var(--primary-color);
  margin-bottom: 60px;
  }

  p,
  input,
  select,
  textarea,
  button {
    font-family: var(--main-font);
    color: var(--primary-color);
    font-size: 16px;
  }

  h1 {
    font-size: 2.5rem;
    font-weight: 900;
  }

  a {
    text-decoration: none;
  }

  .large-p, .large-a {
    font-size: 22px;
  }

  .secondary-font-color {
    color: var(--secondary-color);
  }
`;
