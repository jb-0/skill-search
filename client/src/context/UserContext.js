import React, { createContext, useEffect, useState } from 'react';

export const UserContext = createContext({
  isLoggedIn: undefined,
});

export const UserProvider = (props) => {
  const [userData, setUserData] = useState({
    isLoggedIn: undefined,
  });

  async function checkLoggedIn() {
    const res = await fetch(
      '/api/user/isloggedin',
      {
        method: 'GET',
        headers: {
          Accept: 'application/json',
          'Content-Type': 'application/json',
          frontend: 'react-frontend',
          credentials: 'include'
        },
      }
    );
    
    // Fail safe, if for any reason we cannot reach the server or the request is bad and a non 200
    // is returned we switch to false to ensure private routes remain private.
    if (res.status === 200) {
      setUserData({ isLoggedIn: await res.json() });
    } else {
      setUserData({ isLoggedIn: false });
    }
  }

  useEffect(() => {
    checkLoggedIn();
  }, []);

  return <UserContext.Provider value={ userData }>{props.children}</UserContext.Provider>;
};


