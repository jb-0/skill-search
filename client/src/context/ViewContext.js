import React, { createContext, useState, useLayoutEffect } from 'react';

export const ViewContext = createContext({
  width: window.innerWidth,
  height: window.innerHeight,
  device: window.innerWidth <= 480 ? 'Mobile' : 'Desktop'
});

export const ViewProvider = (props) => {
  const [size, setSize] = useState({
    width: window.innerWidth,
    height: window.innerHeight,
    device: window.innerWidth <= 480 ? 'Mobile' : 'Desktop'
  });

  useLayoutEffect(() => {
    function updateSize() {
      setSize({ 
        width: window.innerWidth, 
        height: window.innerHeight, 
        device: window.innerWidth <= 480 ? 'Mobile' : 'Desktop' 
      });
    }

    window.addEventListener('resize', updateSize);
    updateSize();

    return () => window.removeEventListener('resize', updateSize);
  }, []);
  return <ViewContext.Provider value={ size }>{props.children}</ViewContext.Provider>;
};
