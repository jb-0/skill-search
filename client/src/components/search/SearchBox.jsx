import React, { useContext, useRef } from 'react';
import './SearchBox.css';
import './SearchSuggestion';
import { ViewContext } from '../../context/ViewContext';
import { SearchContext } from '../../context/SearchContext';

// TEMP TERMS DATA TILL MOVED INTO DB
const permittedTerms = [
  'Angular',
  'Babel',
  'CSS',
  'C++',
  'Django',
  'DOM',
  'Express',
  'Go',
  'HTML',
  'Java',
  'JavaScript',
  'Jest',
  'Kotlin',
  'MongoDB',
  'Node',
  'Objective-C',
  'Python',
  'React',
  'React testing library',
  'Ruby on Rails',
  'SQL'
];

function Search(props) {
  const size = useContext(ViewContext);
  const [search, setSearch]  = useContext(SearchContext);
  const searchBoxRef = useRef();

  // Handle changes to the text box and provide suggested terms
  function handleTextBoxUpdates(event) {
    const target = event.target.name;
    const value = event.target.value;

    let suggested;
    if (value.length > 0) {
      suggested = permittedTerms.filter(
        (word) =>
          word.substring(0, value.length).toLowerCase() === value.toLowerCase()
      );
    } else {
      suggested = [];
    }

    setSearch((previousValues) => {
      return { ...previousValues, [target]: value, suggestedTerms: suggested };
    });
  }

  // Scroll down on focus to ensure user see's suggestions and selected items (especially on mobile)
  function handleOnFocus(){
    searchBoxRef.current.scrollIntoView({ behavior: 'smooth' });
  }

  return (
    <div
      className={
        size.device === 'Desktop'
          ? 'search-container'
          : 'search-container-mobile search-container'
      }
    >
      <input
        ref={searchBoxRef}
        onChange={handleTextBoxUpdates}
        onFocus={handleOnFocus}
        type="text"
        id="searchInputText"
        name="searchInputText"
        value={search.searchInputText}
        placeholder="Start typing a skill.."
        autoComplete="off"
      ></input>
    </div>
  );
}

export default Search;
