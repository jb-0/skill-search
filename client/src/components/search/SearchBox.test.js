import '@testing-library/jest-dom/extend-expect';
import React from 'react';
import { unmountComponentAtNode } from 'react-dom';
import { render, screen } from '@testing-library/react';
import renderer from 'react-test-renderer';
import { act } from 'react-dom/test-utils';
import userEvent from '@testing-library/user-event';

import SearchBox from './SearchBox';
import { SearchProvider } from '../../context/SearchContext';

describe('SearchBox component', () => {
  let container = null;
  beforeEach(() => {
    // setup a DOM element as a render target
    container = document.createElement('div');
    document.body.appendChild(container);

    act(() => {
      render(
        <SearchProvider>
          <SearchBox />
        </SearchProvider>,
        container
      );
    });
  });

  afterEach(() => {
    // cleanup on exiting
    unmountComponentAtNode(container);
    container.remove();
    container = null;
  });

  it('search box is successfully rendered', () => {
    expect(screen.getByRole('textbox')).toBeInTheDocument();
  });

  it('the bound value matches user typed input', async () => {
    // Imitate user typing each letter of Jav and assert it is displayed in textbox
    await userEvent.type(screen.getByRole('textbox'), 'Jav');
    expect(screen.getByDisplayValue('Jav')).toBeInTheDocument();
  });

  it('renders correctly when passed arrays containing suggested terms and search terms', () => {
    const tree = renderer
      .create(
        <SearchProvider>
          <SearchBox />
        </SearchProvider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
