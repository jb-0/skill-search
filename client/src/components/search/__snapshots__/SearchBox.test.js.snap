// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchBox component renders correctly when passed arrays containing suggested terms and search terms 1`] = `
<div
  className="search-container"
>
  <input
    autoComplete="off"
    id="searchInputText"
    name="searchInputText"
    onChange={[Function]}
    type="text"
    value=""
  />
  <div>
    <div
      className="suggested-term-container"
      data-testid="suggested-term-container"
      id="node"
      onClick={[Function]}
    >
      <p
        id="node"
      >
        node
      </p>
    </div>
    <div
      className="suggested-term-container"
      data-testid="suggested-term-container"
      id="sql"
      onClick={[Function]}
    >
      <p
        id="sql"
      >
        sql
      </p>
    </div>
  </div>
  <div>
    <div
      className="search-term-container"
      data-testid="search-term-container"
    >
      <p
        id="node"
        onClick={[Function]}
      >
        x 
        node
      </p>
    </div>
    <div
      className="search-term-container"
      data-testid="search-term-container"
    >
      <p
        id="sql"
        onClick={[Function]}
      >
        x 
        sql
      </p>
    </div>
  </div>
</div>
`;

exports[`SearchBox component renders correctly when passed empty arrays 1`] = `
<div
  className="search-container"
>
  <input
    autoComplete="off"
    id="searchInputText"
    name="searchInputText"
    onChange={[Function]}
    type="text"
    value=""
  />
  <div />
  <div />
</div>
`;
